mesonbuild/modules/java.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2021 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pathlib
+from mesonbuild.mesonlib.universal import File
+from mesonbuild.build import CustomTarget
+from mesonbuild.optinterpreter import permitted_kwargs
+import os
+from mesonbuild.interpreterbase.decorators import FeatureNew, typed_pos_args
+import typing as T
+from . import ExtensionModule, ModuleReturnValue, ModuleState
+from ..interpreter import Interpreter
+from mesonbuild.interpreter.interpreterobjects import FileHolder
+
+class JavaModule(ExtensionModule):
+    @FeatureNew('Java Module', '0.59.0')
+    def __init__(self, interpreter: Interpreter):
+        super().__init__(interpreter)
+        self.methods.update({
+            'generate_native_header': self.generate_native_header,
+        })
+
+    @typed_pos_args('generate_native_header', (str, FileHolder), min_varargs=1, max_varargs=1)
+    # @permitted_kwargs({'package'})
+    def generate_native_header(self, state: ModuleState, args: T.Tuple[T.Union[str, FileHolder]], kwargs: T.Dict[str, str]) -> ModuleReturnValue:
+        assert state.backend
+
+        javac = state.find_program("javac")
+        package = kwargs.get('package', None)
+
+        file = args[0].held_object if isinstance(args[0], FileHolder) else File(False, state.subdir, args[0])
+
+        if package:
+            header = f'{package.replace(".", "_")}_{pathlib.Path(file.fname).stem}.h'
+        else:
+            header = f'{pathlib.Path(file.fname).stem}.h'
+
+        ct_kwargs = {
+            'input': file,
+            'output': header,
+            'command': [
+                *javac.get_command(),
+                '-d',
+                os.path.join(state.backend.build_dir, state.subdir),
```

Comment:
Do we have some sort of trash directory or private dir that I could throw the .class file in?

Reply:
@PRIVATE_DIR@